File: src/main/java/no/nav/syfo/consumer/ws/BehandlendeEnhetConsumer.java
package no.nav.syfo.consumer.ws;

import lombok.extern.slf4j.Slf4j;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.ArbeidsfordelingV1;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.FinnBehandlendeEnhetListeUgyldigInput;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.*;
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.meldinger.WSFinnBehandlendeEnhetListeRequest;
import no.nav.tjeneste.virksomhet.person.v3.binding.HentGeografiskTilknytningPersonIkkeFunnet;
import no.nav.tjeneste.virksomhet.person.v3.binding.HentGeografiskTilknytningSikkerhetsbegrensing;
import no.nav.tjeneste.virksomhet.person.v3.binding.PersonV3;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.GeografiskTilknytning;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.Kodeverdi;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.NorskIdent;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.PersonIdent;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentGeografiskTilknytningRequest;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentGeografiskTilknytningResponse;
import org.springframework.stereotype.Component;

import javax.inject.Inject;

import static java.util.Optional.ofNullable;
import static no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.WSEnhetsstatus.AKTIV;

@Component
@Slf4j
public class BehandlendeEnhetConsumer {

    private final PersonV3 personV3;
    private final ArbeidsfordelingV1 arbeidsfordelingV1;

    @Inject
    public BehandlendeEnhetConsumer(PersonV3 personV3, ArbeidsfordelingV1 arbeidsfordelingV1) {
        this.personV3 = personV3;
        this.arbeidsfordelingV1 = arbeidsfordelingV1;
    }

    public String hentBehandlendeEnhet(String fnr) {
        GeografiskTilknytningDTO geografiskTilknytning = hentGeografiskTilknytning(fnr);

        try {
            return arbeidsfordelingV1.finnBehandlendeEnhetListe(new WSFinnBehandlendeEnhetListeRequest()
                    .withArbeidsfordelingKriterier(new WSArbeidsfordelingKriterier()
                            .withDiskresjonskode(
                                    geografiskTilknytning.diskresjonskode != null
                                            ? new WSDiskresjonskoder().withValue(geografiskTilknytning.diskresjonskode)
                                            : null)
                            .withGeografiskTilknytning(new WSGeografi().withValue(geografiskTilknytning.geografiskTilknytning))
                            .withTema(new WSTema().withValue("SYK"))))
                    .getBehandlendeEnhetListe()
                    .stream()
                    .filter(wsOrganisasjonsenhet -> AKTIV.equals(wsOrganisasjonsenhet.getStatus()))
                    .map(WSOrganisasjonsenhet::getEnhetId)
                    .findFirst()
                    .orElseThrow(() -> new RuntimeException("Fant ingen aktiv enhet for " + geografiskTilknytning.geografiskTilknytning));
        } catch (FinnBehandlendeEnhetListeUgyldigInput e) {
            log.error("Feil ved henting av brukers forvaltningsenhet", e);
            throw new RuntimeException("Feil ved henting av brukers forvaltningsenhet", e);
        } catch (RuntimeException e) {
            log.error("Klarte ikke å hente behandlende enhet!", e);
            throw new RuntimeException(e);
        }
    }

    public GeografiskTilknytningDTO hentGeografiskTilknytning(String fnr) {
        try {
            HentGeografiskTilknytningResponse response = personV3.hentGeografiskTilknytning(new HentGeografiskTilknytningRequest()
                    .withAktoer(new PersonIdent().withIdent(new NorskIdent().withIdent(fnr))));

            return GeografiskTilknytningDTO
                    .builder()
                    .geografiskTilknytning(
                            ofNullable(response.getGeografiskTilknytning())
                                    .map(GeografiskTilknytning::getGeografiskTilknytning)
                                    .orElse(null))
                    .diskresjonskode(
                            ofNullable(response.getDiskresjonskode())
                                    .map(Kodeverdi::getValue)
                                    .orElse(null))
                    .build();
        } catch (HentGeografiskTilknytningSikkerhetsbegrensing | HentGeografiskTilknytningPersonIkkeFunnet e) {
            log.error("Feil ved henting av geografisk tilknytning", e);
            throw new RuntimeException("Feil ved henting av geografisk tilknytning", e);
        }
    }

}

File: src/main/kotlin/no/nav/syfo/consumer/ws/BehandlendeEnhetConsumer.kt
package no.nav.syfo.consumer.ws

import log
import no.nav.helse.arbeidsgiver.integrasjoner.pdl.PdlClient
import no.nav.syfo.behandling.BehandlendeEnhetFeiletException
import no.nav.syfo.behandling.FinnBehandlendeEnhetListeUgyldigInputException
import no.nav.syfo.behandling.IngenAktivEnhetException
import no.nav.syfo.domain.GeografiskTilknytningData
import no.nav.syfo.util.Metrikk
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.binding.ArbeidsfordelingV1
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.binding.FinnBehandlendeEnhetListeUgyldigInput
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.ArbeidsfordelingKriterier
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Diskresjonskoder
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Enhetsstatus
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Geografi
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.informasjon.Tema
import no.nav.tjeneste.virksomhet.arbeidsfordeling.v1.meldinger.FinnBehandlendeEnhetListeRequest

const val SYKEPENGER_UTLAND = "4474"

class BehandlendeEnhetConsumer(
    private val pdlClient: PdlClient,
    private val arbeidsfordelingV1: ArbeidsfordelingV1,
    private val metrikk: Metrikk
)  {

    var log = log()

    fun hentBehandlendeEnhet(fnr: String, uuid: String): String {
        val geografiskTilknytning = hentGeografiskTilknytning(fnr)

        val tema = Tema()
        tema.value = "SYK"

        val disk = Diskresjonskoder()
        disk.value = geografiskTilknytning.diskresjonskode

        val geo = Geografi()
        geo.value = geografiskTilknytning.geografiskTilknytning

        val kriterier = ArbeidsfordelingKriterier()
        kriterier.tema = tema
        kriterier.diskresjonskode = null
        kriterier.geografiskTilknytning = geo


        if (geografiskTilknytning.diskresjonskode != null){
            kriterier.diskresjonskode = disk
        }

        val request = FinnBehandlendeEnhetListeRequest()
        request.arbeidsfordelingKriterier = kriterier

        try {
            val behandlendeEnhet = finnAktivBehandlendeEnhet(request, geografiskTilknytning)
            if (SYKEPENGER_UTLAND == behandlendeEnhet) {
                metrikk.tellInntektsmeldingSykepengerUtland()
            }
            log.info("Fant geografiskTilknytning ${geografiskTilknytning.geografiskTilknytning} med behandlendeEnhet $behandlendeEnhet for inntektsmelding $uuid");
            return behandlendeEnhet

        } catch (e: FinnBehandlendeEnhetListeUgyldigInput) {
            log.error("Feil ved henting av brukers forvaltningsenhet", e)
            throw FinnBehandlendeEnhetListeUgyldigInputException(e)
        } catch (e: RuntimeException) {
            log.error("Klarte ikke å hente behandlende enhet!", e)
            throw BehandlendeEnhetFeiletException(e)
        }
    }

    private fun finnAktivBehandlendeEnhet(
        request: FinnBehandlendeEnhetListeRequest,
        geografiskTilknytning: GeografiskTilknytningData
    ): String {
        val behandlendeEnhet = arbeidsfordelingV1.finnBehandlendeEnhetListe(request)
            .behandlendeEnhetListe
            .stream()
            .filter { wsOrganisasjonsenhet -> Enhetsstatus.AKTIV == wsOrganisasjonsenhet.status }
            .map { it.enhetId }
            .findFirst()
            .orElseThrow { IngenAktivEnhetException(geografiskTilknytning.geografiskTilknytning, null) }
        return behandlendeEnhet
    }


    fun hentGeografiskTilknytning(fnr: String): GeografiskTilknytningData {
        pdlClient.fullPerson(fnr).let {
            return GeografiskTilknytningData(
                geografiskTilknytning = it?.hentGeografiskTilknytning?.hentTilknytning(),
                diskresjonskode = it?.hentPerson?.trekkUtDiskresjonskode()
            )
        }
    }

}
